source/FreeRTOS_DNS_Callback.c

```diff
@@ -157,34 +174,56 @@
     {
         const ListItem_t * pxIterator;
         const ListItem_t * xEnd = listGET_END_MARKER( &xCallbackList );
+        DNSCallback_t * pxKeepCallback;
 
-        vTaskSuspendAll();
+        do
         {
-            for( pxIterator = ( const ListItem_t * ) listGET_NEXT( xEnd );
-                 pxIterator != xEnd; )
+            vTaskSuspendAll();
             {
-                DNSCallback_t * pxCallback = ( ( DNSCallback_t * ) listGET_LIST_ITEM_OWNER( pxIterator ) );
-                /* Move to the next item because we might remove this item */
-                pxIterator = ( const ListItem_t * ) listGET_NEXT( pxIterator );
+                pxKeepCallback = NULL;
 
-                if( ( pvSearchID != NULL ) && ( pvSearchID == pxCallback->pvSearchID ) )
+                for( pxIterator = ( const ListItem_t * ) listGET_NEXT( xEnd );
```

Comment:
Why not remember the `pxIterator` value too?

This would allow us to continue from the last place where we were before we hit the `break` statement; making the code more fast and optimized. (think when the entries we want to remove are around the end of the list)

It is as simple as moving the initialization of `pxIterator` outside all the loops (just before the do-while loop begins).

Reply:
I have thought of doing so, but the problem is that the `List_t` is shared with other tasks.
Possibly, we're OK keeping `pxIterator` and continuing were we were. That looks a lot better.
A user task may call `vDNSSetCallBack()`, which will add a new item at the end of the list.
The only problem I see is this function:
~~~c
        void FreeRTOS_gethostbyname_cancel( void * pvSearchID )
        {
            /* _HT_ Should better become a new API call to have the IP-task remove the callback */
            vDNSCheckCallBack( pvSearchID );
        }
~~~
My initials `_HT_` say that actually the IP-task should remove the callback.
I did so:
~~~c
    /* Let the IP-task remove the callback from the list. */
    IPStackEvent_t xRemoveEvent = { eCancelDNSCallbackEvent, pvSearchID };

    /* Send the network task the appropriate event. */
    xSendEventStructToIPTask( &( xRemoveEvent ), portMAX_DELAY );
~~~
and now it is safe to resume/suspend:
~~~c
    ( void ) xTaskResumeAll();
    {
        pCallbackFunction( pcName, pxCallback->pvSearchID, ulIPAddress );
    }
    vTaskSuspendAll();
~~~